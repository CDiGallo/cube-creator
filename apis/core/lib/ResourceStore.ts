import type StreamClient from 'sparql-http-client/StreamClient'
import { NamedNode } from 'rdf-js'
import cf, { GraphPointer } from 'clownface'
import $rdf from 'rdf-ext'
import ToQuadsTransform from 'rdf-transform-triple-to-quad'
import { CONSTRUCT } from '@tpluscode/sparql-builder'

export interface ResourceStore {
  get(id: string | NamedNode, throwWhenNotFound: boolean): Promise<GraphPointer<NamedNode>>
  put(pointer: GraphPointer): Promise<void>
}

export default class implements ResourceStore {
  private readonly __client: StreamClient

  constructor(client: StreamClient) {
    this.__client = client
  }

  /**
   * Gets all triples from a named graph and returns a graph pointer to a resource
   * identified by the same URI
   *
   * @param id
   */
  async get(id: string | NamedNode): Promise<GraphPointer<NamedNode>> {
    const term = typeof id === 'string' ? $rdf.namedNode(id) : id

    const stream = await CONSTRUCT`?s ?p ?o`
      .WHERE`GRAPH ${term} { ?s ?p ?o }`
      .execute(this.__client.query)

    const dataset = await $rdf.dataset().import(stream)

    return cf({ dataset, term })
  }

  /**
   * Replaces the resource in the triple store by inserting default graph
   * quads from the in-memory dataset into a named graph identified by the
   * graph pointer
   *
   * @param pointer Resource to store
   */
  put(pointer: GraphPointer<NamedNode>): Promise<void> {
    const defaultGraphTriples = $rdf.dataset([
      ...pointer.dataset
        .match(null, null, null, $rdf.defaultGraph()),
    ])

    return this.__client.store.put(defaultGraphTriples.toStream().pipe(new ToQuadsTransform(pointer.term)))
  }
}
