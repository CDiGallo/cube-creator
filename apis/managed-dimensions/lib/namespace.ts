import namespace from '@rdfjs/namespace'
import env from './env'

type Shapes = 'shape/managed-dimension' | 'shape/managed-dimension-term-create' | 'shape/managed-dimension-term-update'

export const shape = namespace<Shapes>(`${env.MANAGED_DIMENSIONS_BASE}`)
