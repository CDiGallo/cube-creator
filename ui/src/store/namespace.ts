import { namespace } from 'vuex-class'

export const api = namespace('api')
export const app = namespace('app')
export const auth = namespace('auth')
export const project = namespace('project')
export const projects = namespace('projects')
export const sharedDimension = namespace('sharedDimension')
export const sharedDimensions = namespace('sharedDimensions')
