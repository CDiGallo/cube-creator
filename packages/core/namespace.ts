import namespace from '@rdf-esm/namespace'
import { NamedNode } from 'rdf-js'

export { shape } from './namespaces/shapes'

type CubeCreatorClass =
  'CubeProject' |
  'ProjectsCollection' |
  'CsvMapping' |
  'EntryPoint' |
  'CSVSource' |
  'CSVSourceCollection' |
  'CreateProjectAction' |
  'UploadCSVAction' |
  'Table'

type CubeCreatorProperty =
  'projectSourceKind' |
  'projects' |
  'csvMapping' |
  'csvSource' |
  'csvSourceCollection' |
  'csvColumnSample' |
  'csvw'

type OtherTerms =
  'dash'

type CubeCreatorNamespace = Record<CubeCreatorClass | CubeCreatorProperty | OtherTerms, NamedNode>

export const cube = namespace('http://ns.bergnet.org/cube/')
export const cc: CubeCreatorNamespace = namespace('https://cube-creator.zazuko.com/vocab#') as any
