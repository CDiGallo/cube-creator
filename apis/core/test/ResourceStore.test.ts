import { describe, it, beforeEach } from 'mocha'
import { expect } from 'chai'
import sinon from 'sinon'
import $rdf from 'rdf-ext'
import { StreamClient } from 'sparql-http-client/StreamClient'
import StreamQuery from 'sparql-http-client/StreamQuery'
import StreamStore from 'sparql-http-client/StreamStore'
import ResourceStore from '../lib/ResourceStore'
import { ex } from './support/namespace'

describe('ResourceStore', () => {
  let client: StreamClient
  let query: sinon.SinonStubbedInstance<StreamQuery>
  let store: sinon.SinonStubbedInstance<StreamStore>

  beforeEach(() => {
    query = sinon.createStubInstance(StreamQuery)
    store = sinon.createStubInstance(StreamStore)

    client = {
      query,
      store: store as any,
    }
  })

  describe('get', () => {
    it('loads resource from sparql endpoint', async () => {
      // given
      const store = new ResourceStore(client)
      query.construct.resolves($rdf.dataset().toStream())

      // when
      await store.get(ex.Foo)

      // then
      expect(client.query.construct).to.have.been.calledOnce
    })

    it('called twice returns same object', async () => {
      // given
      const store = new ResourceStore(client)
      query.construct.resolves($rdf.dataset().toStream())
      const expected = await store.get(ex.Foo)

      // when
      const actual = await store.get(ex.Foo)

      // then
      expect(actual).to.eq(expected)
    })

    it('returns object previously created', async () => {
      // given
      const store = new ResourceStore(client)
      const expected = store.create(ex.Foo)

      // when
      const actual = await store.get(ex.Foo)

      // then
      expect(actual).to.eq(expected)
      expect(client.query.construct).to.not.have.been.called
    })
  })

  describe('create', () => {
    it('throws when called twice', () => {
      // given
      const store = new ResourceStore(client)

      // when
      store.create(ex.Foo)

      // then
      expect(() => store.create(ex.Foo)).to.throw
    })
  })

  describe('save', () => {
    it('stores resources in individual named graphs', async function () {
      // given
      const resources = new ResourceStore(client)
      resources.create(ex.Foo).addOut(ex.foo, 'foo')
      resources.create(ex.Bar).addOut(ex.bar, 'bar')

      // when
      await resources.save()

      // then
      expect(client.store.put).to.have.been.calledOnce
      const savedDataset = await $rdf.dataset().import(store.put.firstCall.args[0])
      expect(savedDataset.toCanonical()).to.matchSnapshot(this)
    })
  })
})
