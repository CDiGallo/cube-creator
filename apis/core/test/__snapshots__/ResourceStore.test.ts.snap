exports["ResourceStore save stores resources in individual named graphs"] = "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\nPREFIX hydra: <http://www.w3.org/ns/hydra/core#>\n\nDROP SILENT GRAPH <http://example.com/Foo>;\nDROP SILENT GRAPH <http://example.com/Bar>;\n\nINSERT DATA {\n  \n  \nGRAPH <http://example.com/Foo> {\n          <http://example.com/Foo> rdf:type hydra:Resource .\n<http://example.com/Foo> <http://example.com/foo> \"foo\" .\n        }\n  \nGRAPH <http://example.com/Bar> {\n          <http://example.com/Bar> rdf:type hydra:Resource .\n<http://example.com/Bar> <http://example.com/bar> \"bar\" .\n        }\n}";

exports["ResourceStore save only saves changed resources"] = "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n\nDROP SILENT GRAPH <baz>;\n\nINSERT DATA {\n  \n  \nGRAPH <baz> {\n          <baz> rdfs:label \"Baz changed\" .\n        }\n}";

exports["ResourceStore delete delete is adding DROP SILENT GRAPH statement and removes from inserts"] = "DROP SILENT GRAPH <http://example.com/Foo>;\n\nINSERT DATA {\n  \n}";

exports["ResourceStore save only deletes graph if all resource triples were removed"] = "DROP SILENT GRAPH <baz>;\n\nINSERT DATA {\n  \n}";

